docs/v3/core/v3.0.rst

```diff
@@ -1396,125 +1399,111 @@ array, the prefix is the empty string.  For a non-root array with hierarchy path
 Let "+" be the string concatenation operator.
 
-.. note::
-
-   Store and implementation can assume that a client will not try to
-   create both an *array* and *group* at the same path, and thus
-   may skip check of existence of a group/array of the same name.
 
 **Create a group**
 
     To create an explicit group at hierarchy path `P`, perform
-    ``set(group_meta_key(P), value)``, where `value` is the
+    ``set(meta_key(P), value)``, where `value` is the
     serialization of a valid group metadata document.
 
-    If `P` is a non-root path then it is **not** necessary to create
-    or check for the existence of metadata documents for groups at any
-    of the ancestor paths of `P`. Creating a group at path `P` implies
+    Creating a group at path `P` implies
     the existence of groups at all ancestor paths of `P`.
 
 **Create an array**
 
     To create an array at hierarchy path `P`, perform
-    ``set(array_meta_key(P), value)``, where `value` is the
-    serialisation of a valid array metadata document.
+    ``set(meta_key(P), value)``, where `value` is the serialisation of a valid
+    array metadata document.
 
-    If `P` is a non-root path then it is **not** necessary to create
-    or check for the existence of metadata documents for groups at any
-    of the ancestor paths of `P`. Creating an array at path `P`
-    implies the existence of groups at all ancestor paths of `P`.
+    Creating an array at path `P` implies the existence of groups at all
+    ancestor paths of `P`.
 
 **Store chunk data in an array**
 
     To store chunk data in an array at path `P` and chunk coordinate (`j`, `i`,
-    ...), perform ``set(data_key(P, j, i, ...), value)``, where
-    `value` is the serialisation of the corresponding chunk, encoded
-    according to the information in the array metadata stored under
-    the key ``array_meta_key(P)``.
+    ...), perform ``set(data_key(P, j, i, ...), value)``, where `value` is the
+    serialisation of the corresponding chunk, encoded according to the
+    information in the array metadata stored under the key ``meta_key(P)``.
 
 **Retrieve chunk data in an array**
 
     To retrieve chunk data in an array at path `P` and chunk coordinate (`i`,
     `j`, ...), perform ``get(data_key(P, j, i, ...), value)``. The returned
-    value is the serialisation of the corresponding chunk, encoded
-    according to the array metadata stored at ``array_meta_key(P)``.
+    value is the serialisation of the corresponding chunk, encoded according to
+    the array metadata stored at ``meta_key(P)``.
 
 **Discover children of a group**
 
     To discover the children of a group at hierarchy path `P`, perform
-    ``list_dir(P + "/")``. Any returned prefix not being ``c`` or
-    starting with ``__`` indicates a child group implied by some
-    descendant group or array.
+    ``list_dir(P + "/")``. Any returned prefix not starting with ``__``
+    indicates a child group implied by some descendant group or array.
```

Comment:
```suggestion
    ``list_dir(P + "/")``. Any returned prefix ``Q`` not starting with ``__``
    indicates a potential child array or group. To determine if the child is an array or group,
    the document ``meta_key(Q)`` must be checked.
```

What I dislike about this is that two sequential operations are required to discover children. On high latency stores, this creates significant overhead.

Does anyone see an alternative algorithm?

Reply:
I think we can assume that all members are in fact children.  If we want to know if a given child is an array or a group, we do have to read the zarr.json metadata, and that is unavoidable.  As we discussed previously, though, it is likely that in cases where you want to know if it is an array or a group, you will also want to know additional details that would require reading the zarr.json metadata anyway.

Reply:
I rephrased this slightly to clarify it:
> To discover the children of a group at hierarchy path `P`, perform
    ``list_dir(P + "/")``. Any returned prefix ``Q`` not starting with ``__``
    indicates a child array or group. To determine whether the child is
    an array or group, the document ``meta_key(Q)`` must be checked.
    
Reading the metadata is only needed when one needs to clarify if it's a group or array.